styx-common/src/main/java/com/spotify/styx/util/ShardedCounter.java

```diff
@@ -0,0 +1,364 @@
+
+  private final Time time;
+  private final Datastore datastore;
+
+  /**
+   * A weakly consistent view of the state in Datastore, refreshed by ShardedCounter on demand.
+   */
+  private final Cache<String, CounterSnapshot> cache = CacheBuilder.newBuilder()
+      .maximumSize(100_000)
+      .build();
+
+  private static class CounterSnapshot {
+
+    final String counterId;
+    final Instant updatedAt;
+    final Long limit;
+    final Map<Integer, Long> shards;
+    final Time time;
+
+    public CounterSnapshot(Datastore datastore, String counterId, Time time) {
+      limit = getLimit(datastore, counterId);
+
+      Map<Integer, Long> fetchedShards = fetchShards(datastore, counterId);
+      if (fetchedShards.size() < NUM_SHARDS) {
+        // The counter probably has not been initialized (so we have empty QueryResults). Also
+        // possible that a prior initialize() crashed halfway, or we got a partial list of shards in
+        // QueryResults due to eventual consistency. In any case, repeated initialization eventually
+        // creates all NUM_SHARDS shards.
+        initialize(datastore, counterId);
+        fetchedShards = fetchShards(datastore, counterId);
+      }
+      this.counterId = counterId;
+      this.time = time;
+      shards = fetchedShards;
+      updatedAt = time.get();
+    }
+
+    /**
+     * Idempotent initialization, so that we don't reset an existing shard to zero - counterId may
+     * have already been initialized and incremented by another process.
+     */
+    private static void initialize(Datastore datastore, String counterId) {
+      for (int i = 0; i < NUM_SHARDS; i++) {
+        final int shardIndex = i;
+        final Key shardKey = datastore.newKeyFactory().setKind(KIND_COUNTER_SHARD)
+            .newKey(counterId + "-" + shardIndex);
+        datastore.runInTransaction(transaction -> {
+          final Entity shard = transaction.get(shardKey);
+          if (shard == null) {
+            transaction.put(Entity.newBuilder(shardKey)
+                                .set(PROPERTY_COUNTER_ID, counterId)
+                                .set(PROPERTY_SHARD_INDEX, shardIndex)
+                                .set(PROPERTY_SHARD_VALUE, 0)
+                                .build());
+          }
+          return null;
+        });
+      }
+    }
+
+    /**
+     * Gets the counter state. Creates it in Datastore if it doesn't exist yet.
+     */
+    private static CounterSnapshot fromDatastore(Datastore datastore, String counterId, Time time) {
+      return new CounterSnapshot(datastore, counterId, time);
+    }
+
+    private static Map<Integer, Long> fetchShards(Datastore datastore, String counterId) {
+      final EntityQuery queryShards = EntityQuery.newEntityQueryBuilder()
+          .setKind(KIND_COUNTER_SHARD)
+          .setFilter(PropertyFilter.eq(PROPERTY_COUNTER_ID, counterId))
+          .setOrderBy(OrderBy.asc("__key__"))
```

Comment:
We can remove this - as you pointed out the lexicographical ordering was of no use at all.